editor/groups_editor.cpp

```diff
@@ -59,776 +61,809 @@ static bool can_edit(Node *p_node, String p_group) {
-			error->popup_centered();
-			return;
-		}
-	}
+	updating_tree = true;
 
-	renamed_group->set_text(0, name); // Spaces trimmed.
+	tree->clear();
 
-	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
-	undo_redo->create_action(TTR("Rename Group"));
+	List<Node::GroupInfo> groups;
+	node->get_groups(&groups);
+	groups.sort_custom<_GroupInfoComparator>();
 
-	List<Node *> nodes;
-	scene_tree->get_nodes_in_group(selected_group, &nodes);
-	bool removed_all = true;
-	for (Node *node : nodes) {
-		if (can_edit(node, selected_group)) {
-			undo_redo->add_do_method(node, "remove_from_group", selected_group);
-			undo_redo->add_undo_method(node, "remove_from_group", name);
-			undo_redo->add_do_method(node, "add_to_group", name, true);
-			undo_redo->add_undo_method(node, "add_to_group", selected_group, true);
-		} else {
-			removed_all = false;
-		}
+	List<StringName> current_groups;
+	for (const Node::GroupInfo &gi : groups) {
+		current_groups.push_back(gi.name);
 	}
 
-	if (!removed_all) {
-		undo_redo->add_do_method(this, "_add_group", selected_group);
-		undo_redo->add_undo_method(this, "_delete_group_item", selected_group);
+	TreeItem *root = tree->create_item();
+
+	TreeItem *local_root = tree->create_item(root);
+	local_root->set_text(0, "Scene Groups");
+	local_root->set_icon(0, get_theme_icon(SNAME("PackedScene"), SNAME("EditorIcons")));
+	local_root->set_custom_bg_color(0, get_theme_color(SNAME("prop_subsection"), SNAME("Editor")));
+	local_root->set_selectable(0, false);
+
+	List<StringName> scene_keys;
+	for (const KeyValue<StringName, bool> &E : scene_groups) {
+		scene_keys.push_back(E.key);
 	}
+	scene_keys.sort_custom<NoCaseComparator>();
 
-	undo_redo->add_do_method(this, "_rename_group_item", selected_group, name);
-	undo_redo->add_undo_method(this, "_rename_group_item", name, selected_group);
-	undo_redo->add_do_method(this, "_group_selected");
-	undo_redo->add_undo_method(this, "_group_selected");
-	undo_redo->add_do_method(this, "emit_signal", "group_edited");
-	undo_redo->add_undo_method(this, "emit_signal", "group_edited");
+	for (const StringName &E : scene_keys) {
+		if (!filter->get_text().is_subsequence_ofn(E)) {
+			continue;
+		}
 
-	undo_redo->commit_action();
-}
+		bool is_editable = can_edit(node, E);
+
+		TreeItem *item = tree->create_item(local_root);
+		item->set_cell_mode(0, is_editable ? TreeItem::CELL_MODE_CHECK : TreeItem::CELL_MODE_STRING);
+		item->set_editable(0, is_editable);
+		item->set_checked(0, current_groups.find(E) != nullptr);
+		item->set_text(0, E);
+		item->set_meta("__local", true);
+		item->set_meta("__name", E);
+		item->set_meta("__description", "");
+		if (!is_editable) {
+			item->set_icon(0, get_theme_icon(SNAME("GuiCheckedDisabled"), SNAME("EditorIcons")));
+		}
+		if (!scene_groups[E]) {
+			item->add_button(0, get_theme_icon(SNAME("Lock"), SNAME("EditorIcons")), -1, true, TTR("This group can't be edited outside a packed scene"));
```

Comment:
```suggestion
			item->add_button(0, get_theme_icon(SNAME("Lock"), SNAME("EditorIcons")), -1, true, TTR("This group belongs to another scene and can't be edited."));
```
This option shows up in inherited scenes too.